In [ ]:
#| default_exp 90-ngame-ep-for-wikiseealso-with-lora

In [ ]:
%load_ext autoreload
%autoreload 2

In [ ]:
#| export
import os,torch, torch.multiprocessing as mp, pickle
from xcai.basics import *
from xcai.models.PPP0XX import DBT010

from peft import (
    LoraConfig, 
    get_peft_model, 
    TaskType,
    PeftModel
)

comet_ml is installed but `COMET_API_KEY` is not set.


In [ ]:
os.environ['WANDB_MODE'] = 'disabled'

In [ ]:
#| export
os.environ['CUDA_VISIBLE_DEVICES'] = '0,1'
os.environ['WANDB_PROJECT']='xc-nlg_90-ngame-ep-for-wikiseealso-with-lora'

In [ ]:
#| export
data_dir = '/home/scai/phd/aiz218323/Projects/XC_NLG/data'

In [ ]:
block = XCBlock.from_cfg(data_dir, 'data', valid_pct=0.001, tfm='xcnlg', tokenizer='distilbert-base-uncased', smp_features=[('lbl2data',1,1)])

In [ ]:
#| export
pkl_dir = '/home/scai/phd/aiz218323/scratch/datasets'
pkl_file = f'{pkl_dir}/processed/wikiseealso_data_distilbert-base-uncased_xcnlg_ngame.pkl'

In [ ]:
with open(pkl_file, 'wb') as file: pickle.dump(block, file)

In [ ]:
#| export
with open(pkl_file, 'rb') as file: block = pickle.load(file)

## Training

In [ ]:
#| export
args = XCLearningArguments(
    output_dir='/home/scai/phd/aiz218323/scratch/outputs/90-ngame-ep-for-wikiseealso-with-lora',
    logging_first_step=True,
    per_device_train_batch_size=800,
    per_device_eval_batch_size=800,
    representation_num_beams=200,
    representation_accumulation_steps=100,
    save_strategy="steps",
    evaluation_strategy="steps",
    eval_steps=5000,
    save_steps=5000,
    save_total_limit=5,
    num_train_epochs=300,
    predict_with_representation=True,
    representation_search_type='BRUTEFORCE',
    adam_epsilon=1e-6,
    warmup_steps=100,
    weight_decay=0.01,
    learning_rate=2e-4,
    group_by_cluster=True,
    num_clustering_warmup_epochs=10,
    num_cluster_update_epochs=5,
    num_cluster_size_update_epochs=25,
    clustering_type='EXPO',
    minimum_cluster_size=2,
    maximum_cluster_size=1600,
    target_indices_key='plbl2data_idx',
    target_pointer_key='plbl2data_data2ptr',
    use_encoder_parallel=True,
    max_grad_norm=None,
    fp16=True,
)

In [ ]:
#| export
metric = PrecRecl(block.n_lbl, block.test.data_lbl_filterer, prop=block.train.dset.data.data_lbl,
                  pk=10, rk=200, rep_pk=[1, 3, 5, 10], rep_rk=[10, 100, 200])

In [ ]:
#| export
bsz = max(args.per_device_train_batch_size, args.per_device_eval_batch_size)*torch.cuda.device_count()

model = DBT010.from_pretrained('sentence-transformers/msmarco-distilbert-base-v4', bsz=bsz, tn_targ=5000, margin=0.3, tau=0.1, 
                               n_negatives=10, apply_softmax=True, use_encoder_parallel=False)
model.init_dr_head()

Some weights of DBT010 were not initialized from the model checkpoint at sentence-transformers/msmarco-distilbert-base-v4 and are newly initialized: ['encoder.dr_layer_norm.bias', 'encoder.dr_layer_norm.weight', 'encoder.dr_projector.bias', 'encoder.dr_projector.weight', 'encoder.dr_transform.bias', 'encoder.dr_transform.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [ ]:
#| export
lora_config = LoraConfig(
    r=8,
    lora_alpha=32,
    lora_dropout=0.05,
    target_modules=["q_lin", "k_lin","v_lin"],
    bias='none',
)

# peft_model = get_peft_model(model,lora_config)    

In [ ]:
def func():
    import pdb; pdb.set_trace()
    return get_peft_model(model,lora_config)
    

In [ ]:
o = func()

> /tmp/ipykernel_18917/3479025385.py(3)func()
      1 def func():
      2     import pdb; pdb.set_trace()
----> 3     return get_peft_model(model,lora_config)
      4 



ipdb>  s


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(133)get_peft_model()
    131 
    132 
--> 133 def get_peft_model(
    134     model: PreTrainedModel,
    135     peft_config: PeftConfig,



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(161)get_peft_model()
    159             the base model
    160     """
--> 161     model_config = getattr(model, "config", {"model_type": "custom"})
    162     if hasattr(model_config, "to_dict"):
    163         model_config = model_config.to_dict()



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(162)get_peft_model()
    160     """
    161     model_config = getattr(model, "config", {"model_type": "custom"})
--> 162     if hasattr(model_config, "to_dict"):
    163         model_config = model_config.to_dict()
    164 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(163)get_peft_model()
    161     model_config = getattr(model, "config", {"model_type": "custom"})
    162     if hasattr(model_config, "to_dict"):
--> 163         model_config = model_config.to_dict()
    164 
    165     peft_config.base_model_name_or_path = model.__dict__.get("name_or_path", None)



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(165)get_peft_model()
    163         model_config = model_config.to_dict()
    164 
--> 165     peft_config.base_model_name_or_path = model.__dict__.get("name_or_path", None)
    166 
    167     if revision is not None:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(167)get_peft_model()
    165     peft_config.base_model_name_or_path = model.__dict__.get("name_or_path", None)
    166 
--> 167     if revision is not None:
    168         if peft_config.revision is not None and peft_config.revision != revision:
    169             warnings.warn(



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(174)get_peft_model()
    172         peft_config.revision = revision
    173 
--> 174     if mixed:
    175         # note: PeftMixedModel does not support autocast_adapter_dtype, so don't pass it
    176         return PeftMixedModel(model, peft_config, adapter_name=adapter_name)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(178)get_peft_model()
    176         return PeftMixedModel(model, peft_config, adapter_name=adapter_name)
    177 
--> 178     if peft_config.task_type not in MODEL_TYPE_TO_PEFT_MODEL_MAPPING.keys() and not peft_config.is_prompt_learning:
    179         return PeftModel(model, peft_config, adapter_name=adapter_name, autocast_adapter_dtype=autocast_adapter_dtype)
    180 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(179)get_peft_model()
    177 
    178     if peft_config.task_type not in MODEL_TYPE_TO_PEFT_MODEL_MAPPING.keys() and not peft_config.is_prompt_learning:
--> 179         return PeftModel(model, peft_config, adapter_name=adapter_name, autocast_adapter_dtype=autocast_adapter_dtype)
    180 
    181     if peft_config.is_prompt_learning:



ipdb>  s


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(132)__init__()
    130     """
    131 
--> 132     def __init__(
    133         self,
    134         model: PreTrainedModel,



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(139)__init__()
    137         autocast_adapter_dtype: bool = True,
    138     ) -> None:
--> 139         super().__init__()
    140         self.modules_to_save = None
    141         self.active_adapter = adapter_name



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(140)__init__()
    138     ) -> None:
    139         super().__init__()
--> 140         self.modules_to_save = None
    141         self.active_adapter = adapter_name
    142         self.peft_type = peft_config.peft_type



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(141)__init__()
    139         super().__init__()
    140         self.modules_to_save = None
--> 141         self.active_adapter = adapter_name
    142         self.peft_type = peft_config.peft_type
    143         # These args are special PEFT arguments that users can pass. They need to be removed before passing them to



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(142)__init__()
    140         self.modules_to_save = None
    141         self.active_adapter = adapter_name
--> 142         self.peft_type = peft_config.peft_type
    143         # These args are special PEFT arguments that users can pass. They need to be removed before passing them to
    144         # forward.



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(145)__init__()
    143         # These args are special PEFT arguments that users can pass. They need to be removed before passing them to
    144         # forward.
--> 145         self.special_peft_forward_args = {"adapter_names"}
    146 
    147         self._is_prompt_learning = peft_config.is_prompt_learning



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(147)__init__()
    145         self.special_peft_forward_args = {"adapter_names"}
    146 
--> 147         self._is_prompt_learning = peft_config.is_prompt_learning
    148         if self._is_prompt_learning:
    149             self._peft_config = {adapter_name: peft_config}



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(148)__init__()
    146 
    147         self._is_prompt_learning = peft_config.is_prompt_learning
--> 148         if self._is_prompt_learning:
    149             self._peft_config = {adapter_name: peft_config}
    150             self.base_model = model



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(153)__init__()
    151             self.add_adapter(adapter_name, peft_config)
    152         else:
--> 153             self._peft_config = None
    154             cls = PEFT_TYPE_TO_MODEL_MAPPING[peft_config.peft_type]
    155             self.base_model = cls(model, {adapter_name: peft_config}, adapter_name)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(154)__init__()
    152         else:
    153             self._peft_config = None
--> 154             cls = PEFT_TYPE_TO_MODEL_MAPPING[peft_config.peft_type]
    155             self.base_model = cls(model, {adapter_name: peft_config}, adapter_name)
    156             self.set_additional_trainable_modules(peft_config, adapter_name)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(155)__init__()
    153             self._peft_config = None
    154             cls = PEFT_TYPE_TO_MODEL_MAPPING[peft_config.peft_type]
--> 155             self.base_model = cls(model, {adapter_name: peft_config}, adapter_name)
    156             self.set_additional_trainable_modules(peft_config, adapter_name)
    157 



ipdb>  s


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(138)__init__()
    136     prefix: str = "lora_"
    137 
--> 138     def __init__(self, model, config, adapter_name) -> None:
    139         super().__init__(model, config, adapter_name)
    140 



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(139)__init__()
    137 
    138     def __init__(self, model, config, adapter_name) -> None:
--> 139         super().__init__(model, config, adapter_name)
    140 
    141     def _check_new_adapter_config(self, config: LoraConfig) -> None:



ipdb>  s


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(146)__init__()
    144     """
    145 
--> 146     def __init__(
    147         self,
    148         model,



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(152)__init__()
    150         adapter_name: str,
    151     ) -> None:
--> 152         super().__init__()
    153 
    154         self.model = model



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(154)__init__()
    152         super().__init__()
    153 
--> 154         self.model = model
    155         self.targeted_module_names: list[str] = []
    156 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(155)__init__()
    153 
    154         self.model = model
--> 155         self.targeted_module_names: list[str] = []
    156 
    157         # For advanced developers, if you want to attach multiple adapters to your



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(159)__init__()
    157         # For advanced developers, if you want to attach multiple adapters to your
    158         # model, just add a `peft_config` dict attribute to your model.
--> 159         if not hasattr(self, "peft_config"):
    160             self.peft_config = {adapter_name: peft_config} if isinstance(peft_config, PeftConfig) else peft_config
    161         else:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(160)__init__()
    158         # model, just add a `peft_config` dict attribute to your model.
    159         if not hasattr(self, "peft_config"):
--> 160             self.peft_config = {adapter_name: peft_config} if isinstance(peft_config, PeftConfig) else peft_config
    161         else:
    162             logger.info(



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(172)__init__()
    170                 self.peft_config.update(peft_config)
    171 
--> 172         self.active_adapter: str | list[str] = adapter_name
    173         self._pre_injection_hook(self.model, self.peft_config[adapter_name], adapter_name)
    174         if peft_config != PeftType.XLORA or peft_config[adapter_name] != PeftType.XLORA:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(173)__init__()
    171 
    172         self.active_adapter: str | list[str] = adapter_name
--> 173         self._pre_injection_hook(self.model, self.peft_config[adapter_name], adapter_name)
    174         if peft_config != PeftType.XLORA or peft_config[adapter_name] != PeftType.XLORA:
    175             self.inject_adapter(self.model, adapter_name)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(174)__init__()
    172         self.active_adapter: str | list[str] = adapter_name
    173         self._pre_injection_hook(self.model, self.peft_config[adapter_name], adapter_name)
--> 174         if peft_config != PeftType.XLORA or peft_config[adapter_name] != PeftType.XLORA:
    175             self.inject_adapter(self.model, adapter_name)
    176 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(175)__init__()
    173         self._pre_injection_hook(self.model, self.peft_config[adapter_name], adapter_name)
    174         if peft_config != PeftType.XLORA or peft_config[adapter_name] != PeftType.XLORA:
--> 175             self.inject_adapter(self.model, adapter_name)
    176 
    177         # Copy the peft_config in the injected model.



ipdb>  s


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(353)__getattr__()
    351         return new_module
    352 
--> 353     def __getattr__(self, name: str):
    354         """Forward missing attributes to the wrapped module."""
    355         try:



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(355)__getattr__()
    353     def __getattr__(self, name: str):
    354         """Forward missing attributes to the wrapped module."""
--> 355         try:
    356             return super().__getattr__(name)  # defer to nn.Module's logic
    357         except AttributeError:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(356)__getattr__()
    354         """Forward missing attributes to the wrapped module."""
    355         try:
--> 356             return super().__getattr__(name)  # defer to nn.Module's logic
    357         except AttributeError:
    358             if name == "model":  # see #1892: prevent infinite recursion if class is not initialized



ipdb>  


--Return--
DBT010(
  (en... )
    )
  )
)
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(356)__getattr__()
    354         """Forward missing attributes to the wrapped module."""
    355         try:
--> 356             return super().__getattr__(name)  # defer to nn.Module's logic
    357         except AttributeError:
    358             if name == "model":  # see #1892: prevent infinite recursion if class is not initialized



ipdb>  


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(366)inject_adapter()
    364         pass
    365 
--> 366     def inject_adapter(self, model: nn.Module, adapter_name: str, autocast_adapter_dtype: bool = True) -> None:
    367         r"""
    368         Creates adapter layers and replaces the target modules with the adapter layers. This method is called under the



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(381)inject_adapter()
    379                 Whether to autocast the adapter dtype. Defaults to `True`.
    380         """
--> 381         peft_config = self.peft_config[adapter_name]
    382         # Note: If possible, all checks should be performed *at the start of this method*.
    383         # This way, we can raise early if something goes wrong, without leaving the model



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(385)inject_adapter()
    383         # This way, we can raise early if something goes wrong, without leaving the model
    384         # in a bad (half-initialized) state.
--> 385         self._check_new_adapter_config(peft_config)
    386 
    387         _check_for_modules_to_save = getattr(peft_config, "modules_to_save", None) is not None



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(387)inject_adapter()
    385         self._check_new_adapter_config(peft_config)
    386 
--> 387         _check_for_modules_to_save = getattr(peft_config, "modules_to_save", None) is not None
    388         _has_modules_to_save = False
    389 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(388)inject_adapter()
    386 
    387         _check_for_modules_to_save = getattr(peft_config, "modules_to_save", None) is not None
--> 388         _has_modules_to_save = False
    389 
    390         model_config = getattr(model, "config", {"model_type": "custom"})



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(390)inject_adapter()
    388         _has_modules_to_save = False
    389 
--> 390         model_config = getattr(model, "config", {"model_type": "custom"})
    391         if hasattr(model_config, "to_dict"):
    392             model_config = model_config.to_dict()



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(391)inject_adapter()
    389 
    390         model_config = getattr(model, "config", {"model_type": "custom"})
--> 391         if hasattr(model_config, "to_dict"):
    392             model_config = model_config.to_dict()
    393 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(392)inject_adapter()
    390         model_config = getattr(model, "config", {"model_type": "custom"})
    391         if hasattr(model_config, "to_dict"):
--> 392             model_config = model_config.to_dict()
    393 
    394         peft_config = self._prepare_adapter_config(peft_config, model_config)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(394)inject_adapter()
    392             model_config = model_config.to_dict()
    393 
--> 394         peft_config = self._prepare_adapter_config(peft_config, model_config)
    395 
    396         self._prepare_model(peft_config, model)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(396)inject_adapter()
    394         peft_config = self._prepare_adapter_config(peft_config, model_config)
    395 
--> 396         self._prepare_model(peft_config, model)
    397         is_target_modules_in_base_model = False
    398         key_list = [key for key, _ in model.named_modules()]



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(397)inject_adapter()
    395 
    396         self._prepare_model(peft_config, model)
--> 397         is_target_modules_in_base_model = False
    398         key_list = [key for key, _ in model.named_modules()]
    399 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(398)inject_adapter()
    396         self._prepare_model(peft_config, model)
    397         is_target_modules_in_base_model = False
--> 398         key_list = [key for key, _ in model.named_modules()]
    399 
    400         if getattr(peft_config, "target_modules", None) == DUMMY_TARGET_MODULES:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(400)inject_adapter()
    398         key_list = [key for key, _ in model.named_modules()]
    399 
--> 400         if getattr(peft_config, "target_modules", None) == DUMMY_TARGET_MODULES:
    401             # dummy adapter, we allow not matching any module
    402             key_list = []



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(406)inject_adapter()
    404 
    405         # update peft_config.target_modules if required
--> 406         peft_config = _maybe_include_all_linear_layers(peft_config, model)
    407 
    408         for key in key_list:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(408)inject_adapter()
    406         peft_config = _maybe_include_all_linear_layers(peft_config, model)
    407 
--> 408         for key in key_list:
    409             # Check for modules_to_save in case
    410             if _check_for_modules_to_save and any(



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(410)inject_adapter()
    408         for key in key_list:
    409             # Check for modules_to_save in case
--> 410             if _check_for_modules_to_save and any(
    411                 key.endswith(f"{module_to_save}") for module_to_save in peft_config.modules_to_save
    412             ):



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(425)inject_adapter()
    423                 continue
    424 
--> 425             if not self._check_target_module_exists(peft_config, key):
    426                 continue
    427 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(426)inject_adapter()
    424 
    425             if not self._check_target_module_exists(peft_config, key):
--> 426                 continue
    427 
    428             self.targeted_module_names.append(key)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(408)inject_adapter()
    406         peft_config = _maybe_include_all_linear_layers(peft_config, model)
    407 
--> 408         for key in key_list:
    409             # Check for modules_to_save in case
    410             if _check_for_modules_to_save and any(



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(410)inject_adapter()
    408         for key in key_list:
    409             # Check for modules_to_save in case
--> 410             if _check_for_modules_to_save and any(
    411                 key.endswith(f"{module_to_save}") for module_to_save in peft_config.modules_to_save
    412             ):



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(425)inject_adapter()
    423                 continue
    424 
--> 425             if not self._check_target_module_exists(peft_config, key):
    426                 continue
    427 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(426)inject_adapter()
    424 
    425             if not self._check_target_module_exists(peft_config, key):
--> 426                 continue
    427 
    428             self.targeted_module_names.append(key)



ipdb>  b 428


Breakpoint 3 at /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py:428


ipdb>  r


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(428)inject_adapter()
    426                 continue
    427 
3-> 428             self.targeted_module_names.append(key)
    429             is_target_modules_in_base_model = True
    430             parent, target, target_name = _get_submodules(model, key)



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(429)inject_adapter()
    427 
3   428             self.targeted_module_names.append(key)
--> 429             is_target_modules_in_base_model = True
    430             parent, target, target_name = _get_submodules(model, key)
    431             self._create_and_replace(peft_config, adapter_name, target, target_name, parent, current_key=key)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(430)inject_adapter()
3   428             self.targeted_module_names.append(key)
    429             is_target_modules_in_base_model = True
--> 430             parent, target, target_name = _get_submodules(model, key)
    431             self._create_and_replace(peft_config, adapter_name, target, target_name, parent, current_key=key)
    432 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(431)inject_adapter()
    429             is_target_modules_in_base_model = True
    430             parent, target, target_name = _get_submodules(model, key)
--> 431             self._create_and_replace(peft_config, adapter_name, target, target_name, parent, current_key=key)
    432 
    433         # Handle X-LoRA case.



ipdb>  s


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(173)_create_and_replace()
    171             replicate_layers(model, peft_config.layer_replication)
    172 
--> 173     def _create_and_replace(
    174         self,
    175         lora_config,



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(182)_create_and_replace()
    180         current_key,
    181     ):
--> 182         if current_key is None:
    183             raise ValueError("Current Key shouldn't be `None`")
    184 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(186)_create_and_replace()
    184 
    185         # Regexp matching - Find key which matches current target_name in patterns provided
--> 186         pattern_keys = list(chain(lora_config.rank_pattern.keys(), lora_config.alpha_pattern.keys()))
    187         target_name_key = next(filter(lambda key: re.match(rf".*\.{key}$", current_key), pattern_keys), current_key)
    188         r = lora_config.rank_pattern.get(target_name_key, lora_config.r)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(187)_create_and_replace()
    185         # Regexp matching - Find key which matches current target_name in patterns provided
    186         pattern_keys = list(chain(lora_config.rank_pattern.keys(), lora_config.alpha_pattern.keys()))
--> 187         target_name_key = next(filter(lambda key: re.match(rf".*\.{key}$", current_key), pattern_keys), current_key)
    188         r = lora_config.rank_pattern.get(target_name_key, lora_config.r)
    189         alpha = lora_config.alpha_pattern.get(target_name_key, lora_config.lora_alpha)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(188)_create_and_replace()
    186         pattern_keys = list(chain(lora_config.rank_pattern.keys(), lora_config.alpha_pattern.keys()))
    187         target_name_key = next(filter(lambda key: re.match(rf".*\.{key}$", current_key), pattern_keys), current_key)
--> 188         r = lora_config.rank_pattern.get(target_name_key, lora_config.r)
    189         alpha = lora_config.alpha_pattern.get(target_name_key, lora_config.lora_alpha)
    190 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(189)_create_and_replace()
    187         target_name_key = next(filter(lambda key: re.match(rf".*\.{key}$", current_key), pattern_keys), current_key)
    188         r = lora_config.rank_pattern.get(target_name_key, lora_config.r)
--> 189         alpha = lora_config.alpha_pattern.get(target_name_key, lora_config.lora_alpha)
    190 
    191         kwargs = {



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(192)_create_and_replace()
    190 
    191         kwargs = {
--> 192             "r": r,
    193             "lora_alpha": alpha,
    194             "lora_dropout": lora_config.lora_dropout,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(193)_create_and_replace()
    191         kwargs = {
    192             "r": r,
--> 193             "lora_alpha": alpha,
    194             "lora_dropout": lora_config.lora_dropout,
    195             "fan_in_fan_out": lora_config.fan_in_fan_out,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(194)_create_and_replace()
    192             "r": r,
    193             "lora_alpha": alpha,
--> 194             "lora_dropout": lora_config.lora_dropout,
    195             "fan_in_fan_out": lora_config.fan_in_fan_out,
    196             "init_lora_weights": lora_config.init_lora_weights,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(195)_create_and_replace()
    193             "lora_alpha": alpha,
    194             "lora_dropout": lora_config.lora_dropout,
--> 195             "fan_in_fan_out": lora_config.fan_in_fan_out,
    196             "init_lora_weights": lora_config.init_lora_weights,
    197             "use_rslora": lora_config.use_rslora,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(196)_create_and_replace()
    194             "lora_dropout": lora_config.lora_dropout,
    195             "fan_in_fan_out": lora_config.fan_in_fan_out,
--> 196             "init_lora_weights": lora_config.init_lora_weights,
    197             "use_rslora": lora_config.use_rslora,
    198             "use_dora": lora_config.use_dora,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(197)_create_and_replace()
    195             "fan_in_fan_out": lora_config.fan_in_fan_out,
    196             "init_lora_weights": lora_config.init_lora_weights,
--> 197             "use_rslora": lora_config.use_rslora,
    198             "use_dora": lora_config.use_dora,
    199             "ephemeral_gpu_offload": lora_config.runtime_config.ephemeral_gpu_offload,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(198)_create_and_replace()
    196             "init_lora_weights": lora_config.init_lora_weights,
    197             "use_rslora": lora_config.use_rslora,
--> 198             "use_dora": lora_config.use_dora,
    199             "ephemeral_gpu_offload": lora_config.runtime_config.ephemeral_gpu_offload,
    200             "loaded_in_8bit": getattr(self.model, "is_loaded_in_8bit", False),



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(199)_create_and_replace()
    197             "use_rslora": lora_config.use_rslora,
    198             "use_dora": lora_config.use_dora,
--> 199             "ephemeral_gpu_offload": lora_config.runtime_config.ephemeral_gpu_offload,
    200             "loaded_in_8bit": getattr(self.model, "is_loaded_in_8bit", False),
    201             "loaded_in_4bit": getattr(self.model, "is_loaded_in_4bit", False),



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(200)_create_and_replace()
    198             "use_dora": lora_config.use_dora,
    199             "ephemeral_gpu_offload": lora_config.runtime_config.ephemeral_gpu_offload,
--> 200             "loaded_in_8bit": getattr(self.model, "is_loaded_in_8bit", False),
    201             "loaded_in_4bit": getattr(self.model, "is_loaded_in_4bit", False),
    202         }



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(201)_create_and_replace()
    199             "ephemeral_gpu_offload": lora_config.runtime_config.ephemeral_gpu_offload,
    200             "loaded_in_8bit": getattr(self.model, "is_loaded_in_8bit", False),
--> 201             "loaded_in_4bit": getattr(self.model, "is_loaded_in_4bit", False),
    202         }
    203 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(191)_create_and_replace()
    189         alpha = lora_config.alpha_pattern.get(target_name_key, lora_config.lora_alpha)
    190 
--> 191         kwargs = {
    192             "r": r,
    193             "lora_alpha": alpha,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(204)_create_and_replace()
    202         }
    203 
--> 204         quant_methods = ["gptq", "aqlm", "awq"]
    205         for quant_method in quant_methods:
    206             quantization_config = get_quantization_config(self.model, method=quant_method)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(205)_create_and_replace()
    203 
    204         quant_methods = ["gptq", "aqlm", "awq"]
--> 205         for quant_method in quant_methods:
    206             quantization_config = get_quantization_config(self.model, method=quant_method)
    207             if quantization_config is not None:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(206)_create_and_replace()
    204         quant_methods = ["gptq", "aqlm", "awq"]
    205         for quant_method in quant_methods:
--> 206             quantization_config = get_quantization_config(self.model, method=quant_method)
    207             if quantization_config is not None:
    208                 kwargs[f"{quant_method}_quantization_config"] = quantization_config



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(207)_create_and_replace()
    205         for quant_method in quant_methods:
    206             quantization_config = get_quantization_config(self.model, method=quant_method)
--> 207             if quantization_config is not None:
    208                 kwargs[f"{quant_method}_quantization_config"] = quantization_config
    209 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(205)_create_and_replace()
    203 
    204         quant_methods = ["gptq", "aqlm", "awq"]
--> 205         for quant_method in quant_methods:
    206             quantization_config = get_quantization_config(self.model, method=quant_method)
    207             if quantization_config is not None:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(206)_create_and_replace()
    204         quant_methods = ["gptq", "aqlm", "awq"]
    205         for quant_method in quant_methods:
--> 206             quantization_config = get_quantization_config(self.model, method=quant_method)
    207             if quantization_config is not None:
    208                 kwargs[f"{quant_method}_quantization_config"] = quantization_config



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(207)_create_and_replace()
    205         for quant_method in quant_methods:
    206             quantization_config = get_quantization_config(self.model, method=quant_method)
--> 207             if quantization_config is not None:
    208                 kwargs[f"{quant_method}_quantization_config"] = quantization_config
    209 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(205)_create_and_replace()
    203 
    204         quant_methods = ["gptq", "aqlm", "awq"]
--> 205         for quant_method in quant_methods:
    206             quantization_config = get_quantization_config(self.model, method=quant_method)
    207             if quantization_config is not None:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(206)_create_and_replace()
    204         quant_methods = ["gptq", "aqlm", "awq"]
    205         for quant_method in quant_methods:
--> 206             quantization_config = get_quantization_config(self.model, method=quant_method)
    207             if quantization_config is not None:
    208                 kwargs[f"{quant_method}_quantization_config"] = quantization_config



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(207)_create_and_replace()
    205         for quant_method in quant_methods:
    206             quantization_config = get_quantization_config(self.model, method=quant_method)
--> 207             if quantization_config is not None:
    208                 kwargs[f"{quant_method}_quantization_config"] = quantization_config
    209 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(205)_create_and_replace()
    203 
    204         quant_methods = ["gptq", "aqlm", "awq"]
--> 205         for quant_method in quant_methods:
    206             quantization_config = get_quantization_config(self.model, method=quant_method)
    207             if quantization_config is not None:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(211)_create_and_replace()
    209 
    210         # note: AdaLoraLayer is a subclass of LoraLayer, we need to exclude it
--> 211         from peft.tuners.adalora import AdaLoraLayer
    212 
    213         if isinstance(target, LoraLayer) and not isinstance(target, AdaLoraLayer):



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(213)_create_and_replace()
    211         from peft.tuners.adalora import AdaLoraLayer
    212 
--> 213         if isinstance(target, LoraLayer) and not isinstance(target, AdaLoraLayer):
    214             target.update_layer(
    215                 adapter_name,



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(224)_create_and_replace()
    222             )
    223         else:
--> 224             new_module = self._create_new_module(lora_config, adapter_name, target, **kwargs)
    225             if adapter_name not in self.active_adapters:
    226                 # adding an additional adapter: it is not automatically trainable



ipdb>  s


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(289)_create_new_module()
    287                 raise NotImplementedError(f"Requested bias: {bias}, is not implemented.")
    288 
--> 289     @staticmethod
    290     def _create_new_module(lora_config, adapter_name, target, **kwargs):
    291         # Collect dispatcher functions to decide what backend to use for the replaced LoRA layer. The order matters,



ipdb>  target


Linear(in_features=768, out_features=768, bias=True)


ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(293)_create_new_module()
    291         # Collect dispatcher functions to decide what backend to use for the replaced LoRA layer. The order matters,
    292         # because the first match is always used. Therefore, the default layers should be checked last.
--> 293         dispatchers = []
    294 
    295         if lora_config._custom_modules:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(295)_create_new_module()
    293         dispatchers = []
    294 
--> 295         if lora_config._custom_modules:
    296             # Experimental custom LoRA module support. Allows users to pass a custom mapping for unsupported layer
    297             # types by impelementing their own LoRA layers.



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(316)_create_new_module()
    314 
    315         # avoid eager bnb import
--> 316         if is_bnb_available():
    317             from .bnb import dispatch_bnb_8bit
    318 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(321)_create_new_module()
    319             dispatchers.append(dispatch_bnb_8bit)
    320 
--> 321         if is_bnb_4bit_available():
    322             from .bnb import dispatch_bnb_4bit
    323 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(326)_create_new_module()
    324             dispatchers.append(dispatch_bnb_4bit)
    325 
--> 326         dispatchers.extend(
    327             [
    328                 dispatch_eetq,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(328)_create_new_module()
    326         dispatchers.extend(
    327             [
--> 328                 dispatch_eetq,
    329                 dispatch_aqlm,
    330                 dispatch_awq,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(329)_create_new_module()
    327             [
    328                 dispatch_eetq,
--> 329                 dispatch_aqlm,
    330                 dispatch_awq,
    331                 dispatch_gptq,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(330)_create_new_module()
    328                 dispatch_eetq,
    329                 dispatch_aqlm,
--> 330                 dispatch_awq,
    331                 dispatch_gptq,
    332                 dispatch_hqq,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(331)_create_new_module()
    329                 dispatch_aqlm,
    330                 dispatch_awq,
--> 331                 dispatch_gptq,
    332                 dispatch_hqq,
    333                 dispatch_megatron,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(332)_create_new_module()
    330                 dispatch_awq,
    331                 dispatch_gptq,
--> 332                 dispatch_hqq,
    333                 dispatch_megatron,
    334                 dispatch_default,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(333)_create_new_module()
    331                 dispatch_gptq,
    332                 dispatch_hqq,
--> 333                 dispatch_megatron,
    334                 dispatch_default,
    335             ]



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(334)_create_new_module()
    332                 dispatch_hqq,
    333                 dispatch_megatron,
--> 334                 dispatch_default,
    335             ]
    336         )



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(327)_create_new_module()
    325 
    326         dispatchers.extend(
--> 327             [
    328                 dispatch_eetq,
    329                 dispatch_aqlm,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(326)_create_new_module()
    324             dispatchers.append(dispatch_bnb_4bit)
    325 
--> 326         dispatchers.extend(
    327             [
    328                 dispatch_eetq,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(338)_create_new_module()
    336         )
    337 
--> 338         new_module = None
    339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(339)_create_new_module()
    337 
    338         new_module = None
--> 339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(340)_create_new_module()
    338         new_module = None
    339         for dispatcher in dispatchers:
--> 340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins
    342                 break



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(341)_create_new_module()
    339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
--> 341             if new_module is not None:  # first match wins
    342                 break
    343 



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(339)_create_new_module()
    337 
    338         new_module = None
--> 339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(340)_create_new_module()
    338         new_module = None
    339         for dispatcher in dispatchers:
--> 340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins
    342                 break



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(341)_create_new_module()
    339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
--> 341             if new_module is not None:  # first match wins
    342                 break
    343 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(339)_create_new_module()
    337 
    338         new_module = None
--> 339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(340)_create_new_module()
    338         new_module = None
    339         for dispatcher in dispatchers:
--> 340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins
    342                 break



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(341)_create_new_module()
    339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
--> 341             if new_module is not None:  # first match wins
    342                 break
    343 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(339)_create_new_module()
    337 
    338         new_module = None
--> 339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(340)_create_new_module()
    338         new_module = None
    339         for dispatcher in dispatchers:
--> 340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins
    342                 break



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(341)_create_new_module()
    339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
--> 341             if new_module is not None:  # first match wins
    342                 break
    343 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(339)_create_new_module()
    337 
    338         new_module = None
--> 339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(340)_create_new_module()
    338         new_module = None
    339         for dispatcher in dispatchers:
--> 340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins
    342                 break



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(341)_create_new_module()
    339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
--> 341             if new_module is not None:  # first match wins
    342                 break
    343 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(339)_create_new_module()
    337 
    338         new_module = None
--> 339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(340)_create_new_module()
    338         new_module = None
    339         for dispatcher in dispatchers:
--> 340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins
    342                 break



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(341)_create_new_module()
    339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
--> 341             if new_module is not None:  # first match wins
    342                 break
    343 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(339)_create_new_module()
    337 
    338         new_module = None
--> 339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(340)_create_new_module()
    338         new_module = None
    339         for dispatcher in dispatchers:
--> 340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins
    342                 break



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(341)_create_new_module()
    339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
--> 341             if new_module is not None:  # first match wins
    342                 break
    343 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(342)_create_new_module()
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins
--> 342                 break
    343 
    344         if new_module is None:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(344)_create_new_module()
    342                 break
    343 
--> 344         if new_module is None:
    345             # no module could be matched
    346             raise ValueError(



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(351)_create_new_module()
    349             )
    350 
--> 351         return new_module
    352 
    353     def __getattr__(self, name: str):



ipdb>  new_module


lora.Linear(
  (base_layer): Linear(in_features=768, out_features=768, bias=True)
  (lora_dropout): ModuleDict(
    (default): Dropout(p=0.05, inplace=False)
  )
  (lora_A): ModuleDict(
    (default): Linear(in_features=768, out_features=8, bias=False)
  )
  (lora_B): ModuleDict(
    (default): Linear(in_features=8, out_features=768, bias=False)
  )
  (lora_embedding_A): ParameterDict()
  (lora_embedding_B): ParameterDict()
  (lora_magnitude_vector): ModuleDict()
)


ipdb>  new_module.base_layer


Linear(in_features=768, out_features=768, bias=True)


ipdb>  new_module.base_layer.data_ptr()


*** AttributeError: 'Linear' object has no attribute 'data_ptr'


ipdb>  new_module.base_layer.weight.data_ptr()


776793664


ipdb>  target.weight.data_ptr()


776793664


ipdb>  n


--Return--
lora.Linear(
...ModuleDict()
)
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(351)_create_new_module()
    349             )
    350 
--> 351         return new_module
    352 
    353     def __getattr__(self, name: str):



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(225)_create_and_replace()
    223         else:
    224             new_module = self._create_new_module(lora_config, adapter_name, target, **kwargs)
--> 225             if adapter_name not in self.active_adapters:
    226                 # adding an additional adapter: it is not automatically trainable
    227                 new_module.requires_grad_(False)



ipdb>  new_module


lora.Linear(
  (base_layer): Linear(in_features=768, out_features=768, bias=True)
  (lora_dropout): ModuleDict(
    (default): Dropout(p=0.05, inplace=False)
  )
  (lora_A): ModuleDict(
    (default): Linear(in_features=768, out_features=8, bias=False)
  )
  (lora_B): ModuleDict(
    (default): Linear(in_features=8, out_features=768, bias=False)
  )
  (lora_embedding_A): ParameterDict()
  (lora_embedding_B): ParameterDict()
  (lora_magnitude_vector): ModuleDict()
)


ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(228)_create_and_replace()
    226                 # adding an additional adapter: it is not automatically trainable
    227                 new_module.requires_grad_(False)
--> 228             self._replace_module(parent, target_name, new_module, target)
    229 
    230     def _replace_module(self, parent, child_name, new_module, child):



ipdb>  s


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(230)_replace_module()
    228             self._replace_module(parent, target_name, new_module, target)
    229 
--> 230     def _replace_module(self, parent, child_name, new_module, child):
    231         setattr(parent, child_name, new_module)
    232         # It's not necessary to set requires_grad here, as that is handled by



ipdb>  nn


<module 'torch.nn' from '/scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg/lib/python3.9/site-packages/torch/nn/__init__.py'>


ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(231)_replace_module()
    229 
    230     def _replace_module(self, parent, child_name, new_module, child):
--> 231         setattr(parent, child_name, new_module)
    232         # It's not necessary to set requires_grad here, as that is handled by
    233         # _mark_only_adapters_as_trainable



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(236)_replace_module()
    234 
    235         # child layer wraps the original module, unpack it
--> 236         if hasattr(child, "base_layer"):
    237             child = child.base_layer
    238 



ipdb>  l


    231         setattr(parent, child_name, new_module)
    232         # It's not necessary to set requires_grad here, as that is handled by
    233         # _mark_only_adapters_as_trainable
    234 
    235         # child layer wraps the original module, unpack it
--> 236         if hasattr(child, "base_layer"):
    237             child = child.base_layer
    238 
    239         if not hasattr(new_module, "base_layer"):
    240             if hasattr(new_module, "W_q"):  # HQQ
    241                 new_module.W_q = child.W_q



ipdb>  child


Linear(in_features=768, out_features=768, bias=True)


ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(239)_replace_module()
    237             child = child.base_layer
    238 
--> 239         if not hasattr(new_module, "base_layer"):
    240             if hasattr(new_module, "W_q"):  # HQQ
    241                 new_module.W_q = child.W_q



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(247)_replace_module()
    245                 new_module.bias = child.bias
    246 
--> 247         if getattr(child, "state", None) is not None:
    248             if hasattr(new_module, "base_layer"):
    249                 new_module.base_layer.state = child.state



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(255)_replace_module()
    253 
    254         # dispatch to correct device
--> 255         for name, module in new_module.named_modules():
    256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (



ipdb>  self.prefix


'lora_'


ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(256)_replace_module()
    254         # dispatch to correct device
    255         for name, module in new_module.named_modules():
--> 256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (
    258                     child.qweight



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(255)_replace_module()
    253 
    254         # dispatch to correct device
--> 255         for name, module in new_module.named_modules():
    256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(256)_replace_module()
    254         # dispatch to correct device
    255         for name, module in new_module.named_modules():
--> 256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (
    258                     child.qweight



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(255)_replace_module()
    253 
    254         # dispatch to correct device
--> 255         for name, module in new_module.named_modules():
    256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(256)_replace_module()
    254         # dispatch to correct device
    255         for name, module in new_module.named_modules():
--> 256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (
    258                     child.qweight



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(259)_replace_module()
    257                 weight = (
    258                     child.qweight
--> 259                     if hasattr(child, "qweight")
    260                     else child.W_q
    261                     if hasattr(child, "W_q")



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(258)_replace_module()
    256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (
--> 258                     child.qweight
    259                     if hasattr(child, "qweight")
    260                     else child.W_q



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(261)_replace_module()
    259                     if hasattr(child, "qweight")
    260                     else child.W_q
--> 261                     if hasattr(child, "W_q")
    262                     else child.weight
    263                     if hasattr(child, "weight")



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(260)_replace_module()
    258                     child.qweight
    259                     if hasattr(child, "qweight")
--> 260                     else child.W_q
    261                     if hasattr(child, "W_q")
    262                     else child.weight



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(263)_replace_module()
    261                     if hasattr(child, "W_q")
    262                     else child.weight
--> 263                     if hasattr(child, "weight")
    264                     else next(child.parameters())
    265                 )



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(262)_replace_module()
    260                     else child.W_q
    261                     if hasattr(child, "W_q")
--> 262                     else child.weight
    263                     if hasattr(child, "weight")
    264                     else next(child.parameters())



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(257)_replace_module()
    255         for name, module in new_module.named_modules():
    256             if (self.prefix in name) or ("ranknum" in name):
--> 257                 weight = (
    258                     child.qweight
    259                     if hasattr(child, "qweight")



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(266)_replace_module()
    264                     else next(child.parameters())
    265                 )
--> 266                 module.to(weight.device)
    267 
    268     def _mark_only_adapters_as_trainable(self, model: nn.Module) -> None:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(255)_replace_module()
    253 
    254         # dispatch to correct device
--> 255         for name, module in new_module.named_modules():
    256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(256)_replace_module()
    254         # dispatch to correct device
    255         for name, module in new_module.named_modules():
--> 256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (
    258                     child.qweight



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(259)_replace_module()
    257                 weight = (
    258                     child.qweight
--> 259                     if hasattr(child, "qweight")
    260                     else child.W_q
    261                     if hasattr(child, "W_q")



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(258)_replace_module()
    256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (
--> 258                     child.qweight
    259                     if hasattr(child, "qweight")
    260                     else child.W_q



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(261)_replace_module()
    259                     if hasattr(child, "qweight")
    260                     else child.W_q
--> 261                     if hasattr(child, "W_q")
    262                     else child.weight
    263                     if hasattr(child, "weight")



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(260)_replace_module()
    258                     child.qweight
    259                     if hasattr(child, "qweight")
--> 260                     else child.W_q
    261                     if hasattr(child, "W_q")
    262                     else child.weight



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(263)_replace_module()
    261                     if hasattr(child, "W_q")
    262                     else child.weight
--> 263                     if hasattr(child, "weight")
    264                     else next(child.parameters())
    265                 )



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(262)_replace_module()
    260                     else child.W_q
    261                     if hasattr(child, "W_q")
--> 262                     else child.weight
    263                     if hasattr(child, "weight")
    264                     else next(child.parameters())



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(257)_replace_module()
    255         for name, module in new_module.named_modules():
    256             if (self.prefix in name) or ("ranknum" in name):
--> 257                 weight = (
    258                     child.qweight
    259                     if hasattr(child, "qweight")



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(266)_replace_module()
    264                     else next(child.parameters())
    265                 )
--> 266                 module.to(weight.device)
    267 
    268     def _mark_only_adapters_as_trainable(self, model: nn.Module) -> None:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(255)_replace_module()
    253 
    254         # dispatch to correct device
--> 255         for name, module in new_module.named_modules():
    256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(256)_replace_module()
    254         # dispatch to correct device
    255         for name, module in new_module.named_modules():
--> 256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (
    258                     child.qweight



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(259)_replace_module()
    257                 weight = (
    258                     child.qweight
--> 259                     if hasattr(child, "qweight")
    260                     else child.W_q
    261                     if hasattr(child, "W_q")



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(258)_replace_module()
    256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (
--> 258                     child.qweight
    259                     if hasattr(child, "qweight")
    260                     else child.W_q



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(261)_replace_module()
    259                     if hasattr(child, "qweight")
    260                     else child.W_q
--> 261                     if hasattr(child, "W_q")
    262                     else child.weight
    263                     if hasattr(child, "weight")



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(260)_replace_module()
    258                     child.qweight
    259                     if hasattr(child, "qweight")
--> 260                     else child.W_q
    261                     if hasattr(child, "W_q")
    262                     else child.weight



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(263)_replace_module()
    261                     if hasattr(child, "W_q")
    262                     else child.weight
--> 263                     if hasattr(child, "weight")
    264                     else next(child.parameters())
    265                 )



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(262)_replace_module()
    260                     else child.W_q
    261                     if hasattr(child, "W_q")
--> 262                     else child.weight
    263                     if hasattr(child, "weight")
    264                     else next(child.parameters())



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(257)_replace_module()
    255         for name, module in new_module.named_modules():
    256             if (self.prefix in name) or ("ranknum" in name):
--> 257                 weight = (
    258                     child.qweight
    259                     if hasattr(child, "qweight")



ipdb>  r


--Return--
None
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(255)_replace_module()
    253 
    254         # dispatch to correct device
--> 255         for name, module in new_module.named_modules():
    256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (



ipdb>  n


--Return--
None
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(228)_create_and_replace()
    226                 # adding an additional adapter: it is not automatically trainable
    227                 new_module.requires_grad_(False)
--> 228             self._replace_module(parent, target_name, new_module, target)
    229 
    230     def _replace_module(self, parent, child_name, new_module, child):



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(408)inject_adapter()
    406         peft_config = _maybe_include_all_linear_layers(peft_config, model)
    407 
--> 408         for key in key_list:
    409             # Check for modules_to_save in case
    410             if _check_for_modules_to_save and any(



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(410)inject_adapter()
    408         for key in key_list:
    409             # Check for modules_to_save in case
--> 410             if _check_for_modules_to_save and any(
    411                 key.endswith(f"{module_to_save}") for module_to_save in peft_config.modules_to_save
    412             ):



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(425)inject_adapter()
    423                 continue
    424 
--> 425             if not self._check_target_module_exists(peft_config, key):
    426                 continue
    427 



ipdb>  r


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(428)inject_adapter()
    426                 continue
    427 
3-> 428             self.targeted_module_names.append(key)
    429             is_target_modules_in_base_model = True
    430             parent, target, target_name = _get_submodules(model, key)



ipdb>  b


Num Type         Disp Enb   Where
1   breakpoint   keep yes   at /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py:428
	breakpoint already hit 4 times
2   breakpoint   keep yes   at /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py:428
3   breakpoint   keep yes   at /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py:428


ipdb>  disable 1


Disabled breakpoint 1 at /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py:428


ipdb>  disable 2


Disabled breakpoint 2 at /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py:428


ipdb>  disable 3


Disabled breakpoint 3 at /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py:428


ipdb>  r


--Return--
None
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(451)inject_adapter()
    449                     p.requires_grad = False
    450 
--> 451         if _has_modules_to_save:
    452             if not hasattr(model, "modules_to_save"):
    453                 model.modules_to_save = set(peft_config.modules_to_save)



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(178)__init__()
    176 
    177         # Copy the peft_config in the injected model.
--> 178         self.model.peft_config = self.peft_config
    179 
    180     @property



ipdb>  


--Return--
None
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(178)__init__()
    176 
    177         # Copy the peft_config in the injected model.
--> 178         self.model.peft_config = self.peft_config
    179 
    180     @property



ipdb>  


--Return--
None
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(139)__init__()
    137 
    138     def __init__(self, model, config, adapter_name) -> None:
--> 139         super().__init__(model, config, adapter_name)
    140 
    141     def _check_new_adapter_config(self, config: LoraConfig) -> None:



ipdb>  


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg/lib/python3.9/site-packages/torch/nn/modules/module.py(1690)__setattr__()
   1688         raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
   1689 
-> 1690     def __setattr__(self, name: str, value: Union[Tensor, 'Module']) -> None:
   1691         def remove_from(*dicts_or_sets):
   1692             for d in dicts_or_sets:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg/lib/python3.9/site-packages/torch/nn/modules/module.py(1691)__setattr__()
   1689 
   1690     def __setattr__(self, name: str, value: Union[Tensor, 'Module']) -> None:
-> 1691         def remove_from(*dicts_or_sets):
   1692             for d in dicts_or_sets:
   1693                 if name in d:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg/lib/python3.9/site-packages/torch/nn/modules/module.py(1699)__setattr__()
   1697                         d.discard(name)
   1698 
-> 1699         params = self.__dict__.get('_parameters')
   1700         if isinstance(value, Parameter):
   1701             if params is None:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg/lib/python3.9/site-packages/torch/nn/modules/module.py(1700)__setattr__()
   1698 
   1699         params = self.__dict__.get('_parameters')
-> 1700         if isinstance(value, Parameter):
   1701             if params is None:
   1702                 raise AttributeError(



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg/lib/python3.9/site-packages/torch/nn/modules/module.py(1706)__setattr__()
   1704             remove_from(self.__dict__, self._buffers, self._modules, self._non_persistent_buffers_set)
   1705             self.register_parameter(name, value)
-> 1706         elif params is not None and name in params:
   1707             if value is not None:
   1708                 raise TypeError(f"cannot assign '{torch.typename(value)}' as parameter '{name}' "



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg/lib/python3.9/site-packages/torch/nn/modules/module.py(1713)__setattr__()
   1711             self.register_parameter(name, value)
   1712         else:
-> 1713             modules = self.__dict__.get('_modules')
   1714             if isinstance(value, Module):
   1715                 if modules is None:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg/lib/python3.9/site-packages/torch/nn/modules/module.py(1714)__setattr__()
   1712         else:
   1713             modules = self.__dict__.get('_modules')
-> 1714             if isinstance(value, Module):
   1715                 if modules is None:
   1716                     raise AttributeError(



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg/lib/python3.9/site-packages/torch/nn/modules/module.py(1715)__setattr__()
   1713             modules = self.__dict__.get('_modules')
   1714             if isinstance(value, Module):
-> 1715                 if modules is None:
   1716                     raise AttributeError(
   1717                         "cannot assign module before Module.__init__() call")



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg/lib/python3.9/site-packages/torch/nn/modules/module.py(1718)__setattr__()
   1716                     raise AttributeError(
   1717                         "cannot assign module before Module.__init__() call")
-> 1718                 remove_from(self.__dict__, self._parameters, self._buffers, self._non_persistent_buffers_set)
   1719                 for hook in _global_module_registration_hooks.values():
   1720                     output = hook(self, name, value)



ipdb>  c


    [... skipped 1 hidden frame]

    [... skipped 1 hidden frame]

    [... skipped 1 hidden frame]

    [... skipped 1 hidden frame]



/scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py:428

self.inject_adapter(self.model, adapter_name)

In [ ]:
for n,p in peft_model.named_parameters():
    if p.requires_grad: print(n)

base_model.model.encoder.distilbert.transformer.layer.0.attention.q_lin.lora_A.default.weight
base_model.model.encoder.distilbert.transformer.layer.0.attention.q_lin.lora_B.default.weight
base_model.model.encoder.distilbert.transformer.layer.0.attention.k_lin.lora_A.default.weight
base_model.model.encoder.distilbert.transformer.layer.0.attention.k_lin.lora_B.default.weight
base_model.model.encoder.distilbert.transformer.layer.0.attention.v_lin.lora_A.default.weight
base_model.model.encoder.distilbert.transformer.layer.0.attention.v_lin.lora_B.default.weight
base_model.model.encoder.distilbert.transformer.layer.1.attention.q_lin.lora_A.default.weight
base_model.model.encoder.distilbert.transformer.layer.1.attention.q_lin.lora_B.default.weight
base_model.model.encoder.distilbert.transformer.layer.1.attention.k_lin.lora_A.default.weight
base_model.model.encoder.distilbert.transformer.layer.1.attention.k_lin.lora_B.default.weight
base_model.model.encoder.distilbert.transformer.layer.1.atte

In [ ]:
#| export
for p in peft_model.parameters():
    p.requires_grad_(True)

In [ ]:
peft_model.base_model._mark_only_adapters_as_trainable(peft_model)

In [ ]:
#| export
learn = XCLearner(
    model=peft_model, 
    args=args,
    train_dataset=block.train.dset,
    eval_dataset=block.test.dset,
    data_collator=block.collator,
    compute_metrics=metric,
)

/scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg/lib/python3.9/site-packages/accelerate/accelerator.py:432: FutureWarning: Passing the following arguments to `Accelerator` is deprecated and will be removed in version 1.0 of Accelerate: dict_keys(['dispatch_batches', 'split_batches', 'even_batches', 'use_seedable_sampler']). Please pass an `accelerate.DataLoaderConfiguration` instead: 
dataloader_config = DataLoaderConfiguration(dispatch_batches=None, split_batches=False, even_batches=True, use_seedable_sampler=True)
  warnings.warn(
Detected kernel version 3.10.0, which is below the recommended minimum of 5.5.0; this can cause the process to hang. It is recommended to upgrade the kernel to the minimum version or higher.


In [ ]:
learn.train()

Failed to detect the name of this notebook, you can set it manually with the WANDB_NOTEBOOK_NAME environment variable to enable code saving.


  0%|          | 0/347 [00:00<?, ?it/s]

In [ ]:
#| export
if __name__ == '__main__':
    mp.freeze_support()
    learn.train()

## Prediction

In [ ]:
args = XCLearningArguments(
    output_dir='/home/scai/phd/aiz218323/scratch/outputs/59-ngame-ep-for-wikiseealso-with-cls-for-dr-1-1',
    logging_first_step=True,
    per_device_train_batch_size=800,
    per_device_eval_batch_size=800,
    representation_num_beams=200,
    representation_accumulation_steps=100,
    predict_with_representation=True,
    representation_search_type='BRUTEFORCE',
    target_indices_key='plbl2data_idx',
    target_pointer_key='plbl2data_data2ptr',
    use_encoder_parallel=True,
    fp16=True,
)

In [ ]:
output_dir = f"/home/scai/phd/aiz218323/scratch/outputs/{os.path.basename(args.output_dir)}"
mname = f'{output_dir}/{os.path.basename(get_best_model(output_dir))}'

In [ ]:
bsz = max(args.per_device_train_batch_size, args.per_device_eval_batch_size)*torch.cuda.device_count()

model = DBT009.from_pretrained('sentence-transformers/msmarco-distilbert-base-v4', bsz=bsz, tn_targ=5000, margin=0.3, tau=0.1,
                               n_negatives=10, apply_softmax=True, use_encoder_parallel=True)

Some weights of DBT009 were not initialized from the model checkpoint at sentence-transformers/msmarco-distilbert-base-v4 and are newly initialized: ['encoder.dr_layer_norm.bias', 'encoder.dr_layer_norm.weight', 'encoder.dr_projector.bias', 'encoder.dr_projector.weight', 'encoder.dr_transform.bias', 'encoder.dr_transform.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [ ]:
from safetensors import safe_open

model_weight_file = f'{mname}/model.safetensors'

model_weights = {}
with safe_open(model_weight_file, framework="pt") as file:
    for k in file.keys(): model_weights[k] = file.get_tensor(k)
        

In [ ]:
model.load_state_dict(model_weights, strict=False)

_IncompatibleKeys(missing_keys=['distilbert.embeddings.word_embeddings.weight', 'distilbert.embeddings.position_embeddings.weight', 'distilbert.embeddings.LayerNorm.weight', 'distilbert.embeddings.LayerNorm.bias', 'distilbert.transformer.layer.0.attention.q_lin.weight', 'distilbert.transformer.layer.0.attention.q_lin.bias', 'distilbert.transformer.layer.0.attention.k_lin.weight', 'distilbert.transformer.layer.0.attention.k_lin.bias', 'distilbert.transformer.layer.0.attention.v_lin.weight', 'distilbert.transformer.layer.0.attention.v_lin.bias', 'distilbert.transformer.layer.0.attention.out_lin.weight', 'distilbert.transformer.layer.0.attention.out_lin.bias', 'distilbert.transformer.layer.0.sa_layer_norm.weight', 'distilbert.transformer.layer.0.sa_layer_norm.bias', 'distilbert.transformer.layer.0.ffn.lin1.weight', 'distilbert.transformer.layer.0.ffn.lin1.bias', 'distilbert.transformer.layer.0.ffn.lin2.weight', 'distilbert.transformer.layer.0.ffn.lin2.bias', 'distilbert.transformer.layer.

In [ ]:
metric = PrecRecl(block.n_lbl, block.test.data_lbl_filterer, prop=block.train.dset.data.data_lbl,
                  pk=10, rk=200, rep_pk=[1, 3, 5, 10], rep_rk=[10, 100, 200])

In [ ]:
metric = PrecRecl(block.n_lbl, block.train.data_lbl_filterer, prop=block.train.dset.data.data_lbl,
                  pk=10, rk=200, rep_pk=[1, 3, 5, 10], rep_rk=[10, 100, 200])

In [ ]:
learn = XCLearner(
    model=model, 
    args=args,
    train_dataset=block.train.dset,
    eval_dataset=block.test.dset,
    data_collator=block.collator,
    compute_metrics=metric,
)

/scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg/lib/python3.9/site-packages/accelerate/accelerator.py:432: FutureWarning: Passing the following arguments to `Accelerator` is deprecated and will be removed in version 1.0 of Accelerate: dict_keys(['dispatch_batches', 'split_batches', 'even_batches', 'use_seedable_sampler']). Please pass an `accelerate.DataLoaderConfiguration` instead: 
dataloader_config = DataLoaderConfiguration(dispatch_batches=None, split_batches=False, even_batches=True, use_seedable_sampler=True)
  warnings.warn(
Detected kernel version 3.10.0, which is below the recommended minimum of 5.5.0; this can cause the process to hang. It is recommended to upgrade the kernel to the minimum version or higher.


In [ ]:
o = learn.predict(block.test.dset)

  0%|          | 0/391 [00:00<?, ?it/s]

/scratch/scai/phd/aiz218323/Projects/xcai/xcai/losses.py:22: UserWarning: Sparse CSR tensor support is in beta state. If you miss a functionality in the sparse tensor support, please submit a feature request to https://github.com/pytorch/pytorch/issues. (Triggered internally at ../aten/src/ATen/SparseCsrTensorImpl.cpp:53.)
  return torch.sparse_csr_tensor(data_ptr, data_idx, scores, device=data_ptr.device)


/scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg/lib/python3.9/site-packages/scipy/sparse/_index.py:145: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.
  self._set_arrayXarray(i, j, x)


In [ ]:
display_metric(o.metrics)

P@1      P@3      P@5    P@10      N@1     N@3      N@5     N@10  \
0  30.9078  20.5532  15.5551  9.8319  30.9078  30.737  31.7741  33.5844   

    PSP@1    PSP@3    PSP@5  PSP@10   PSN@1    PSN@3    PSN@5   PSN@10  \
0  25.343  27.0596  29.0996  33.186  25.343  27.5471  29.1813  31.2599   

      R@10    R@100    R@200   loss   runtime  samples_per_second  \
0  38.2247  51.9416  55.1189  0.025  620.9983             285.854   

   steps_per_second  
0             0.357

In [ ]:
pred_dir = f"{mname}/predictions/"
os.makedirs(pred_dir, exist_ok=True)

with open(f'{pred_dir}/test_predictions.pkl', 'wb') as file: pickle.dump(o, file)

In [ ]:
o = learn.predict(block.train.dset)

  0%|          | 0/391 [00:00<?, ?it/s]

In [ ]:
display_metric(o.metrics)

P@1      P@3     P@5     P@10      N@1     N@3     N@5     N@10  \
0  92.2863  50.2929  33.625  18.1342  92.2863  93.644  94.357  95.1143   

     PSP@1   PSP@3    PSP@5   PSP@10    PSN@1    PSN@3    PSN@5   PSN@10  \
0  82.7769  87.365  89.3606  92.3762  82.7769  89.3253  91.1272  92.4036   

      R@10    R@100    R@200    loss   runtime  samples_per_second  \
0  97.6791  99.4445  99.5756  0.0011  352.5323             1966.01   

   steps_per_second  
0             2.459

In [ ]:
pred_dir = f"{mname}/predictions/"
os.makedirs(pred_dir, exist_ok=True)

with open(f'{pred_dir}/train_predictions.pkl', 'wb') as file: pickle.dump(o, file)